### DISCUSSION TITLE
what decides docker pull's extraction speed?

### DISCUSSION THREAD

**Discussion by minshi123** [moby/moby#45722] [Q&A]
As far as I know, CPU/IO has an impact on the extraction performance, including using pigz...
When I use different machines but with the same configuration(i.e. same CPU, same IO speed),
they show different results(instead of the expectation: the same speed)...
**Here  I missed something that decides the docker pull speed~?**

PS:
One machine has been running online for long time.
Another machine is a new machine
They all belongs to VM.

**Comment by AkihiroSuda** [comment]
> When I use different machines but with the same configuration(i.e. same CPU, same IO speed),
they show different results(instead of the expectation: the same speed)...

With pigz or without?

**Comment by minshi123** [comment]
> > When I use different machines but with the same configuration(i.e. same CPU, same IO speed),
> > they show different results(instead of the expectation: the same speed)...
> 
> With pigz or without?

using pigz, AFAIK, it doesn't produce any performance difference.. so all machines currently did not install pigz..we are doing the further evaluation about the pigz's actual action.

**Comment by minshi123** [comment]
BTW,   how could I know  my installed pigz does work? I try to start the **Debug** mode, but I cannot see the concrete daemon log corresponding the code section log:
https://github.com/moby/moby/blob/89382f2f20745b9e63bed6c066f104980dff4396/pkg/archive/archive.go#L32-L39

**Comment by minshi123** [comment]
PS: using overlay2 instead of devicemapper, the docker pull speed is apparently faster.... why~? (eg. when pulling node, using devicemapper costs 25s+, but overlay2 costs 16s+

**Comment by thaJeztah** [comment]
Are you running device mapper with `loop-lvm` or `direct-lvm` ? https://docs.docker.com/storage/storagedriver/device-mapper-driver/#device-mapper-and-docker-performance. Performance with `loop-lvm` will definitely be slow.

Generally, I'd recommend using overlay2 (device mapper has also been marked deprecated, although it's not yet actively removed)

**Comment by jobcespedes** [comment]
Also, when storage path is a mount point backed by nfs or iscsi volume, extraction is very slow (both pigz and gzip).
